### PR TITLE
[trainer] feat: Add Torchtitan as alternative training engine

### PR DESCRIPTION
**Pull request by acisseJZhong**
### What does this PR do?
Integrate Torchtitan as a new training engine in Verl. This PR implements the basic APIs needed by Torchtitan Engine, and tested SFT trainer in verl (qwen3 0.6b):
- Torchtitan Engine matches exactly with FSDP engine for SFT trainer
- `use_remove_padding=True` matches `use_remove_padding=False`
- TP/SP and FSDP work with both varlen and flex attention; numerics match with single process. 

**Relevant PRs:**
- Torchtitan side changes in https://github.com/pytorch/torchtitan/pull/2333. 
- RFC for engine interfaces https://github.com/volcengine/verl/issues/1371
- Training engine interface design https://github.com/volcengine/verl/pull/1977 
- Add Veomini Engine https://github.com/verl-project/verl/pull/4072

**Todos:**
See roadmap here: https://github.com/verl-project/verl/issues/5306
- [ ] enable parallelism: enable PP, EP, CP
- [ ] make Torchtitan Engine work with RL trainer 
- [ ] test multimodal input(ref: https://github.com/verl-project/verl/pull/4492/changes)

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test
```
MODEL_ID=Qwen/Qwen3-0.6B BACKEND=torchtitan bash tests/special_e2e/sft/run_sft_engine.sh
MODEL_ID=Qwen/Qwen3-0.6B BACKEND=fsdp bash tests/special_e2e/sft/run_sft_engine.sh
```
`use_remove_padding=True`
image
`use_remove_padding=False`
image

Test TP and FSDP
image

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/utils/torch_functional.py [hunk 1]
```diff
@@ -743,6 +573,8 @@ def get_cosine_schedule_with_warmup(
     assert init_lr_ratio >= 0 and init_lr_ratio <= 1.0
 
     def lr_lambda(current_step):
+        # # 0-indexed step, hence + 1 adjustments
+        current_step += 1
```
Comment (acisseJZhong) [inline verl/utils/torch_functional.py]: see lr discussions here https://github.com/pytorch/torchtitan/pull/2333#discussion_r2783835080

File: tests/special_e2e/sft/run_sft_engine.sh [hunk 2]
```diff
@@ -88,6 +93,27 @@ MEGATRON_ENGINE_CONFIG="\
     engine.context_parallel_size=${CP_SIZE}
     engine.use_mbridge=True"
 
+TORCHTITAN_ENGINE_CONFIG="\
+    engine=${backend} \
+    model=torchtitan_model \
+    model.attn_type=varlen \
+    model.hf_assets_path=${MODEL_PATH}
```
[the remark this answers is not in the document]
Comment (wuxibin89) [reply tests/special_e2e/sft/run_sft_engine.sh]: @acisseJZhong Please resolve this comment.

File: verl/trainer/config/model/torchtitan_model.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# Format checks enforced on CI:
+# 1. Comments must appear above each field.
+# 2. There must be a blank line between each field.
+# 3. Inline comments (after a field on the same line) are not allowed.
+# 4. Indentation level is respected for nested fields.
+
+_target_: verl.workers.config.TorchtitanModelConfig
```
Comment (wuxibin89) [inline verl/trainer/config/model/torchtitan_model.yaml]: Why not reuse `verl/trainer/config/model/hf_model.yaml`?

File: verl/workers/config/model.py [hunk 4]
```diff
@@ -206,3 +159,47 @@ def __post_init__(self):
 
     def get_processor(self):
         return self.processor if self.processor is not None else self.tokenizer
+
+
+@dataclass
+class TorchtitanModelConfig(BaseConfig):
```
Comment (wuxibin89) [inline verl/workers/config/model.py]: Please reuse `HFModelConfig`, add additional fields for torchtitan if needed.

File: verl/trainer/config/optim/torchtitan.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+# Target class for this configuration
+_target_: verl.workers.config.TorchtitanOptimizerConfig
```
Comment (wuxibin89) [inline verl/trainer/config/optim/torchtitan.yaml:2]: Is it possible to reuse `verl/trainer/config/optim/fsdp.yaml`?

File: verl/workers/engine/torchtitan/transformer_impl.py [hunk 6]
```diff
@@ -0,0 +1,671 @@
...
+        """Advance learning rate scheduler."""
+        self.lr_scheduler.step()
+        lr = self.lr_scheduler.schedulers[0].get_last_lr()[0]
+        return lr
+
+    def to(self, device: str, model: bool = True, optimizer: bool = True, grad: bool = True):
+        """Move model and/or optimizer to CPU or GPU."""
+        super().to(device=device, model=model, optimizer=optimizer, grad=grad)
+
+        if self.engine_config.forward_only:
```
Comment (wuxibin89) [inline verl/workers/engine/torchtitan/transformer_impl.py:407]: Does torchtitan share same load/offload mechanism with fsdp backend? In fsdp backend, we use `CPUOffloadPolicy` for `forward_only` reference model, I don't see any cpu offload policy here.

File: tests/special_e2e/sft/run_sft_engine.sh [hunk 7]
```diff
@@ -97,6 +122,10 @@ elif [ "$backend" = "veomni" ]; then
     ENGINE_CONFIG="$VEOMNI_ENGINE_CONFIG"
     echo "Using veomni engine"
     exp_name=gsm8k-${backend}-sp${SP_SIZE}-fsdp${FSDP_SIZE}-pad-${PAD_MODE}-use_remove_padding-${USE_REMOVE_PADDING}-mode-${mode}
+elif [ "$backend" = "torchtitan" ]; then
```
Comment (wuxibin89) [inline tests/special_e2e/sft/run_sft_engine.sh:124]: Please verify different parallelism in `tests/special_e2e/sft/test_sft_engine_all.sh`

File: verl/workers/engine/torchtitan/transformer_impl.py [hunk 8]
```diff
@@ -0,0 +1,671 @@
...
+        # Get train_spec and directly override model_args before Trainer init
+        train_spec = train_spec_module.get_train_spec(self.model_config.name)
+        model_args = train_spec.model_args.get(self.model_config.flavor)
+        if model_args is not None:
+            if hasattr(model_args, "attn_type"):
+                model_args.attn_type = self.model_config.attn_type
+            if hasattr(model_args, "attn_mask_type"):
+                model_args.attn_mask_type = self.model_config.attn_mask_type
+
+        model = Model(
```
Comment (wuxibin89) [inline verl/workers/engine/torchtitan/transformer_impl.py:130]: Is there a model support list in torchtitan? or does it support any huggingface model?

File: verl/trainer/config/engine/torchtitan.yaml [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+entropy_from_logits_with_chunking: false
+
+# Whether to use entropy checkpointing
+entropy_checkpointing: false
+
+# Data parallel size (FSDP group size)
+data_parallel_size: 1
+
+# Data parallel replicate size
+data_parallel_replicate_size: 1
```
Comment (wuxibin89) [inline verl/trainer/config/engine/torchtitan.yaml:35]: Is there any document explain these parallelism?

**Comment by wuxibin89** [comment]
@acisseJZhong Since there're quite some works to do, please open an issue to track torchtitan integration roadmap. https://github.com/verl-project/verl/issues/4880

File: verl/workers/config/model.py (hunk 4 above)
Comment (acisseJZhong) [reply verl/workers/config/model.py]: maybe let `TorchtitanModelConfig` inherit from `HFModelConfig` is a better way? as those fields are likely not used for other training engines/Torchtitan Engine specific.

File: verl/trainer/config/optim/torchtitan.yaml (hunk 5 above)
Comment (acisseJZhong) [reply verl/trainer/config/optim/torchtitan.yaml:2]: I would prefer the separation, as each engine has it's own config which inherits from `OptimizerConfig`. Technically we could merge those two configs together by adding fields to `FSDPOptimizerConfig`, but it's less clear separation? If we do so, it makes less sense to call it `FSDPOptimizerConfig` anyways?

File: verl/workers/engine/torchtitan/transformer_impl.py (hunk 6 above)
Comment (acisseJZhong) [reply verl/workers/engine/torchtitan/transformer_impl.py:407]: nice catch, in Titan it's passed through here https://github.com/pytorch/torchtitan/blob/27930cb612df4d6ebafdb21909749a9694bc167e/torchtitan/config/job_config.py#L270. Will update this in next commit.

File: tests/special_e2e/sft/run_sft_engine.sh (hunk 7 above)
Comment (acisseJZhong) [reply tests/special_e2e/sft/run_sft_engine.sh:124]: sounds good. I will incorporate TP/SP with this PR, for other parallelism, will have separate PRs.

File: verl/workers/engine/torchtitan/transformer_impl.py (hunk 8 above)
Comment (acisseJZhong) [reply verl/workers/engine/torchtitan/transformer_impl.py:130]: this is the model list we supported https://github.com/pytorch/torchtitan/tree/main/torchtitan/models. Right now it doesn't support HF models, but should be relatively easy if user wants to add one. 

We are also actively working on adding VLMs. cc @shuhuayu @tianyu-l

File: verl/trainer/config/engine/torchtitan.yaml (hunk 9 above)
Comment (acisseJZhong) [reply verl/trainer/config/engine/torchtitan.yaml:35]: https://github.com/pytorch/torchtitan/blob/27930cb612df4d6ebafdb21909749a9694bc167e/torchtitan/config/job_config.py#L312

**Comment by acisseJZhong** [comment]
created roadmap here https://github.com/verl-project/verl/issues/5306, @wuxibin89 please feel free to add any items I might be missing.

File: verl/workers/engine/torchtitan/transformer_impl.py (hunk 8 above)
Comment (tianyu-l) [reply verl/workers/engine/torchtitan/transformer_impl.py:130]: We have a model backend aiming to support transformers models out-of-box.
https://github.com/pytorch/torchtitan/tree/main/torchtitan/experiments/transformers_modeling_backend

File: tests/special_e2e/sft/test_sft_engine_all.sh [hunk 10]
```diff
@@ -37,6 +37,14 @@ BACKEND=megatron TP_SIZE=2 PP_SIZE=2 VPP_SIZE=${VPP_SIZE} CP_SIZE=2 NUM_GPUS=8 b
 echo "run with tp2 pp2 vpp2 cp2 num_gpus8 mode=ray"
 BACKEND=megatron TP_SIZE=2 PP_SIZE=2 VPP_SIZE=${VPP_SIZE} CP_SIZE=2 NUM_GPUS=8 mode=ray bash tests/special_e2e/sft/run_sft_engine.sh
 
+# test with torchtitan fsdp=1
+echo "run with tp1 pp1 cp1 fsdp2 num_gpus2"
+BACKEND=torchtitan TP_SIZE=1 PP_SIZE=1 CP_SIZE=1 FSDP_SIZE=2 NUM_GPUS=2 bash tests/special_e2e/sft/run_sft_engine.sh
```
Comment (wuxibin89) [inline tests/special_e2e/sft/test_sft_engine_all.sh]: Well...it breaks NPU and VLM ci, I think we better temporary disable it in ci once we're ready. Checkout ci workflows:
```
.github/workflows/e2e_sft_llm_ascend.yml
.github/workflows/e2e_sft_llm.yml
.github/workflows/e2e_sft_vlm.yml
```

**Comment by wuxibin89** [comment]
sanity check failed:
```bash
python3 tests/special_sanity/check_device_api_usage.py --directory ./verl
```

File: verl/trainer/config/model/torchtitan_model.yaml [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# Format checks enforced on CI:
```
Comment (vermouth1992) [inline verl/trainer/config/model/torchtitan_model.yaml]: Please use verl/trainer/config/model/hf_model.yaml. All the model should start from huggingface

File: verl/utils/torch_functional.py (hunk 1 above)
Comment (vermouth1992) [reply verl/utils/torch_functional.py]: Can we still aligned with megatron if we made this changes?

Comment (vermouth1992) [reply verl/utils/torch_functional.py]: If not, then shall we make this configurable?

**Review by vermouth1992 — CHANGES_REQUESTED** [review]

File: verl/trainer/config/model/torchtitan_model.yaml (hunk 11 above)
Comment (wuxibin89) [reply verl/trainer/config/model/torchtitan_model.yaml]: I think we can add backend specific fields in `hf_model.yaml`, e.g:
```yaml

torchtitan:
  name: qwen3
  flavor: "0.6B"
```

File: tests/special_e2e/sft/test_sft_engine_all.sh (hunk 10 above)
Comment (acisseJZhong) [reply tests/special_e2e/sft/test_sft_engine_all.sh]: ah it's because I didnt pip install torchtitan in the CI flow. let me install and see if it passes.

Comment (acisseJZhong) [reply tests/special_e2e/sft/test_sft_engine_all.sh]: seems CI needs to manual approval after I added `pip install torchtitan`. I will disable torchtitan run in CI for now, we can add back once we have everything including RL ready.

File: verl/trainer/config/model/torchtitan_model.yaml (hunk 11 above)
Comment (vermouth1992) [reply verl/trainer/config/model/torchtitan_model.yaml]: This is not desired. We should strictly start from huggingface naming and checkpoint because this is where people create their models

Comment (acisseJZhong) [reply verl/trainer/config/model/torchtitan_model.yaml]: the `name` and `flavor` fields are required by Titan to get corresponding train spec and model args. It does 1-1 mapping to hugging face models. See more here https://github.com/pytorch/torchtitan/blob/fde830de29c34c55b4cdc0209ac51f5b8084244e/torchtitan/models/llama3/__init__.py#L51

@vermouth1992 @wuxibin89  let me know if you have better ideas, but for now I think what the user need to do is just explicitly pass in `name` and `flavor`. It still aligns with the HF naming and ckpt.

**Comment by acisseJZhong** [comment]
two failing CI test seems irrelevant to this PR, will add titan engine CI after RL trainer is enabled.

File: verl/trainer/config/model/hf_model.yaml [hunk 12]
```diff
@@ -95,3 +95,19 @@ mtp:
 
   method: mtp
   num_speculative_tokens: 1
+
+# Torchtitan backend configuration
+# Only used when engine backend is set to "torchtitan"
+torchtitan:
```
Comment (vermouth1992) [inline verl/trainer/config/model/hf_model.yaml]: This is still not desirable. All the models including names and flavors must start from a single huggingface folder. We can introduce a general model_implementation dict so that users can write `attn_type` and `attn_mask_type` inside this sub-config

Comment (acisseJZhong) [reply verl/trainer/config/model/hf_model.yaml]: added a helper function to derive model `name` and `flavor` from hf config, and get rid of `attn_mask_type` since it's not used. For `attn_type`, I moved it to `TorchtitanEngineConfig` since it's more torchtitan specific field(I don't want other training engine to have this field). Please let me know if you have different opinions @vermouth1992
